### PR TITLE
LPS-73721

### PR DESCRIPTION
**Pull request by jeyvison**
cc: @pcwhite

### CODE REVIEW

**Comment by jeyvison** [comment]
ci:test

**Comment by jeyvison** [comment]
ci:test

**Comment by jeyvison** [comment]
ci:stop

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java [hunk 1]
```diff
@@ -245,6 +246,36 @@ public int getWorkflowDefinitionCount(WorkflowDefinition workflowDefinition)
 			workflowDefinition.getName());
 	}
 
+	public String getWorkflowDefinitionDuplicationTitle(
```
Comment (inacionery) [inline portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: @jeyvison, `getDuplicateTitle` doesn't sound better?

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp [hunk 2]
```diff
@@ -231,9 +239,31 @@ renderResponse.setTitle((workflowDefinition == null) ? LanguageUtil.get(request,
...
+<c:if test="<%= workflowDefinition != null %>">
+	<div class="hide" id="<%= randomId %>titleInputLocalized">
+		<aui:col>
+			<aui:field-wrapper label="title">
+				<liferay-ui:input-localized name="title" xml="<%= duplicationTitle %>" />
+			</aui:field-wrapper>
+		</aui:col>
+
+		<aui:col>
+			<liferay-ui:message key="copy-does-not-include-revisions" translateArguments="<%= false %>" />
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: @jeyvison, `translateArguments="<%= false %>"` in this case is not necessary.

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp [hunk 3]
```diff
@@ -231,9 +239,31 @@ renderResponse.setTitle((workflowDefinition == null) ? LanguageUtil.get(request,
...
+		<aui:input name="content" type="hidden" value="<%= workflowDefinition.getContent() %>" />
+		<aui:input name="duplicatedDefinitionTitle" type="hidden" value="<%= workflowDefinition.getTitle(LanguageUtil.getLanguageId(request)) %>" />
+		<aui:input name="name" type="hidden" value="<%= PortalUUIDUtil.generate() %>" />
+	</div>
+</c:if>
+
 <aui:script use="aui-ace-editor,liferay-xml-formatter,liferay-workflow-web">
 	var STR_VALUE = 'value';
 
+	var title = '<liferay-ui:message key="duplicate-workflow" translateArguments="<%= false %>"/>';
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: @jeyvison, `translateArguments="<%= false %>"` in this case is not necessary.

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp [hunk 4]
```diff
@@ -17,11 +17,20 @@
 <%@ include file="/definition/init.jsp" %>
 
 <%
+String randomId = workflowDefinitionDisplayContext.getWorkflowDefinitionRandomId();
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: @jeyvison, please use it directly, like this https://github.com/liferay/liferay-portal/blob/master/portal-web/docroot/html/taglib/ui/page_iterator/start.jsp#L20

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp [hunk 5]
```diff
@@ -59,6 +68,12 @@ WorkflowDefinition workflowDefinition = (WorkflowDefinition)row.getObject();
 				message="unpublish"
 				url="<%= deactivateWorkflowDefinitionURL %>"
 			/>
+
+			<liferay-ui:icon
+				id='<%= "duplicate" + String.valueOf(workflowDefinition.getName()) %>'
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: @jeyvison, `String.valueOf` is necessary?

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp [hunk 6]
```diff
@@ -74,4 +89,30 @@ WorkflowDefinition workflowDefinition = (WorkflowDefinition)row.getObject();
...
+
+<div class="hide" id="<%= randomId %>titleInputLocalized">
+	<aui:col>
+		<aui:field-wrapper label="title">
+			<liferay-ui:input-localized name="title" xml="<%= duplicationTitle %>" />
+		</aui:field-wrapper>
+	</aui:col>
+
+	<aui:col>
+		<liferay-ui:message key="copy-does-not-include-revisions" translateArguments="<%= false %>" />
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: @jeyvison, translateArguments="<%= false %>" in this case is not necessary.

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp [hunk 7]
```diff
@@ -74,4 +89,30 @@ WorkflowDefinition workflowDefinition = (WorkflowDefinition)row.getObject();
...
+	</aui:col>
+</div>
+
+<div class="hide" id="<%= randomId %>contentInput">
+	<aui:input name="name" type="hidden" value="<%= PortalUUIDUtil.generate() %>" />
+	<aui:input name="content" type="hidden" value="<%= workflowDefinition.getContent() %>" />
+	<aui:input name="duplicatedDefinitionTitle" type="hidden" value="<%= workflowDefinition.getTitle(LanguageUtil.getLanguageId(request)) %>" />
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: @jeyvison, we should use the default locale `LocaleUtil.toLanguageId(LocaleUtil.getDefault());` please extract it to a variable.

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js [hunk 8]
```diff
@@ -2,6 +2,88 @@ AUI.add(
...
 		var WorkflowWeb = {
+			confirmBeforeDuplicateDialog: function(event, actionUrl, title, randomId, content) {
+				var instance = this;
+
+				var form = A.Node.create('<form />');
+
+				form.setAttribute('action', actionUrl);
+				form.setAttribute('method', 'POST');
+
+				var titleInput = A.one('#' + randomId + 'titleInputLocalized');
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: @jeyvison, please move it to the first use.

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js [hunk 9]
```diff
@@ -2,6 +2,88 @@ AUI.add(
...
+
+				var contentInput = A.one('#' + randomId + 'contentInput');
```
Comment (inacionery) [inline portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: @jeyvison, please move it to the first use.

File: portal-workflow-lang/src/main/resources/content/Language.properties [hunk 10]
```diff
@@ -1,12 +1,21 @@
...
 action-undone=Action undone.
 activity-description=Activity Description
 add-new-x=Add New {0}
 an-error-occurred-while-retrieving-the-list-of-instances=An error occurred while retrieving the list of publication requests. Please contact an administrator.
 configure-assignments=Configure Assignments
+copy-does-not-include-revisions=This copy does not include the revision history.
+copy-of=Copy of
 default-display-view=Default Display View
 definition-imported-sucessfully=Definition imported successfully.
-delete-workflow-question=Delete workflow?
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-lang/src/main/resources/content/Language.properties]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 1 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java [hunk 11]
```diff
@@ -245,6 +246,36 @@ public int getWorkflowDefinitionCount(WorkflowDefinition workflowDefinition)
...
+		WorkflowDefinition workflowDefinition) {
+
+		if ((workflowDefinition == null) ||
+			Validator.isNull(workflowDefinition.getContent())) {
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java [hunk 12]
```diff
@@ -245,6 +246,36 @@ public int getWorkflowDefinitionCount(WorkflowDefinition workflowDefinition)
...
+
+			return StringPool.BLANK;
+		}
+
+		ResourceBundle resourceBundle =
+			_resourceBundleLoader.loadResourceBundle(
+				_workflowDefinitionRequestHelper.getLocale());
+
+		String languageId = LocalizationUtil.getDefaultLanguageId(
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java [hunk 13]
```diff
@@ -245,6 +246,36 @@ public int getWorkflowDefinitionCount(WorkflowDefinition workflowDefinition)
...
+			workflowDefinition.getContent());
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java [hunk 14]
```diff
@@ -245,6 +246,36 @@ public int getWorkflowDefinitionCount(WorkflowDefinition workflowDefinition)
...
+
+		String newTitle =
+			LanguageUtil.get(resourceBundle, "copy-of") + StringPool.SPACE +
+				workflowDefinition.getTitle(languageId);
+
+		String updatedTitle = LocalizationUtil.updateLocalization(
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/portlet/action/DuplicateWorkflowDefinitionMVCActionCommand.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	service = MVCActionCommand.class
+)
+public class DuplicateWorkflowDefinitionMVCActionCommand
+	extends UpdateWorkflowDefinitionMVCActionCommand {
+
+	@Override
+	protected void addSuccessMessage(
+		ActionRequest actionRequest, ActionResponse actionResponse) {
+
+		ResourceBundle resourceBundle = resourceBundleLoader.loadResourceBundle(
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/portlet/action/DuplicateWorkflowDefinitionMVCActionCommand.java]: Wont fix

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp [hunk 16]
```diff
@@ -17,6 +17,7 @@
 <%@ include file="/definition/init.jsp" %>
 
 <%
+String randomId = workflowDefinitionDisplayContext.getWorkflowDefinitionRandomId();
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp (hunk 2 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp (hunk 3 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 4 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 5 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 6 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 7 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js (hunk 8 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/portlet/configuration/icon/DuplicateDefinitionPortletConfigurationIcon.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+		"path=/definition/edit_workflow_definition.jsp"
+	},
+	service = PortletConfigurationIcon.class
+)
+public class DuplicateDefinitionPortletConfigurationIcon
+	extends BasePortletConfigurationIcon {
+
+	@Override
+	public String getMessage(PortletRequest portletRequest) {
+		ResourceBundle resourceBundle =
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/portlet/configuration/icon/DuplicateDefinitionPortletConfigurationIcon.java]: wont fix

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js (hunk 9 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: done

File: portal-workflow-lang/src/main/resources/content/Language.properties [hunk 18]
```diff
@@ -1,12 +1,21 @@
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-lang/src/main/resources/content/Language.properties]: done

File: portal-workflow-lang/src/main/resources/content/Language.properties (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-lang/src/main/resources/content/Language.properties]: already fixed

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 1 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/java/com/liferay/portal/workflow/web/internal/display/context/WorkflowDefinitionDisplayContext.java]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp (hunk 16 above)
[the remark this answers is not in the document]
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp (hunk 2 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp (hunk 3 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/edit_workflow_definition.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 4 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 5 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp (hunk 6 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/definition/workflow_definition_action.jsp]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js (hunk 8 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: done

File: portal-workflow-web/src/main/resources/META-INF/resources/js/main.js (hunk 9 above)
Comment (jeyvison) [reply portal-workflow-web/src/main/resources/META-INF/resources/js/main.js]: done

**Comment by inacionery** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by inacionery** [comment]
Forwarded to https://github.com/jeyvison/com-liferay-portal-workflow/pull/10

**Pull request closed without merging**